tools/win/win.1
.\" Dd	distance to space vertically before a "display"
.\" These are what n/troff use for interparagraph distance
.\"-------
.if t .nr Dd .4v
.if n .nr Dd 1v
.\"-------
.\" Ds	begin a display, indented .5 inches from the surrounding text.
.\"
.\" Note that uses of Ds and De may NOT be nested.
.\"-------
.de Ds
.sp \\n(Ddu
.RS \\$1
.nf
..
.\"-------
.\" De	end a display (no trailing vertical spacing)
.\"-------
.de De
.fi
.RE
.in
..
.\" @(#) win %R%.%L% of %D%
.TH win 1 "%R%.%L% of %D%"
.SH NAME
win \- Terminal emulator
.SH SYNOPSIS
.B win
.RB [ \-v ]
.RB [ -t
.IB name ]
.RI [ command
.RI [ argument
.IR ... ]] 
.I address
.SH OPTIONS
.TP
.B \-v
If the
.B \-v
option is present,
.I win
prints its version number and date and exits.
.TP
.BI \-t name
If the
.B \-t
option is present, win uses
.I +name
as the final part of the tag name of its window.
.SH DESCRIPTION
.I win
must be able to connect to a running 
.IR wily (1).
.I win
attaches to a window.
If the
.B \-t
option is present, the final part of the tag name
is
.IR name .
If
.I command
is present, the final part of the tag name is
.IR +command .
Otherwise, the final part of the tag name is
.IR +win .
The first part of the tag name is the current working
directory.
.PP
win runs a program in a child process.
The child process is the leader of a process group
and is connected to
.I win 
by a pseudo-terminal.
If
.I command
is specified, the
child process runs
.Ds
$SHELL -c 'command argument ...'
.De
otherwise it runs
.Ds        
$SHELL -i
.De    
If SHELL is not found in the environment, 
.I /bin/sh
is used.
.PP
TERM=win is placed in the environment of the running program.
.PP        
Output from the running command appears in the window.
The point after the last output from the running command is known
as the output point.
Further output from the running command appears just before the output point.
The output point is initially at the end of the window.
.PP    
Normal wily editing commands work in the window.
When a newline, interrupt character, or end-of-file character 
is created after the output point,
the text between the output point and the last newline,
interrupt character, or end-of-file character in the window 
(inclusive) is passed to the running program as input.
.PP
The interrupt character is control-C and the end-of-file character
is control-D.
The pseudo-terminal initially is configured so that these are
recognized with their normal meanings.
.PP
The B2 commands beginning with the |, <, or > characters
or an upper case letter are executed normally by wily.
Other B2 are first
terminated with a newline, if they are not already, and are
then appended to the buffer (and thereafter passed to the
running program).
.SH RETURNS
.I win
returns zero if it is able to successfully create the 
child process, otherwise it returns
non-zero.
.SH EXAMPLES
Run a terminal emulator within wily
.Ds
win
.De
.PP
Run FTP within wily
.Ds
win ftp
.De
.PP
Run rlogin within wily
.Ds
win -t host.domain rlogin -8 host.domain
.De
.SH SEE ALSO
.IR wily (1)
.IR Tag (1)
.IR Man (1)
.SH BUGS
.I win
doesn't follow changes to the terminal attributes.
In particular, there is no way to stop echoing
or to change its concept of the interrupt and 
eof characters.
.PP
.I rlogin
seems to need -8 and, annoyingly, sets echo
and onlcr.
The latter can be fixed in your .rcrc:
.Ds
if ( ~ $TERM win ) {
    stty -echo -onlcr
}
.De
.PP
.I win
doesn't follow changes to the terminal attributes.
In particular, there is no way to stop echoing
or to change its concept of the interrupt and 
eof characters.
.PP
WEdestroy messages aren't yet passed by wily.
Thus, Del and Delcol can delete the window yet leave
win running.
.PP		
There is a race condition; the user can press newline and
then delete text before win has a chance to read it.
.SH AUTHOR
.I win
was originally written by Gary Capell (gary@cs.su.oz.au).
Alan Watson (alan@oldp.nmsu.edu) rewrote it and added
support for arguments, pseudo-terminals instead of pipes,
and execution of certain B2 commands by wily.
The pseudo-terminal support in 9term, written by
Matty Farrow (matty@cs.su.oz.au), was a great help.


Doc/changes.txt
0.13.12 started this file
* Each line in $WILYTOOLS or $HOME/.wilytools can contain 
either a comment (starts with whitespace or '#') or a regexp,
then a tab, then some tools (i.e. some text) which will be put
into the tag of any window whose label matches the regexp.
The _first_ regexp match wins.
0.13.14
* Scrolling forward works better
* plugged some memory leaks
0.13.15 
fixed silly bug in text_read: Axel Belinfant 
patch from Steve Kotsopoulos - $WMAINTAG and $WCOLTAG 
removed last_escape_select 
newline only special in tag now 
h version 
Version builtin removed 
no longer warps cursor when expanding tile after typing in it 
utfwrite handles short writes -- Bill Trost 
wilytools have newline stripped, b2b1 on Anchor selects line 
changed mouse_expand: clicking _near_ a selection hits the selection 
use view_cut a lot 
removed some functions in keyboard.c 
use view->autoindent to decide to autoindent a view 
Indent builtin 
option -a to set tabsize 
0.13.16 
bugfix with B2B1 
removed Version from default wilytools 
word-delete seems saner to me 
B3 checks for local files before include files 
0.13.17 
splitting data.c up 
tidied a few functions in data.c 
cloned windows share the same tag, modifying label works better 
tag modification seems to be working 
0.13.18 
ctrlu patch -- Bill 
removed some dead code from mouse.c 
renamed globals.h global.h 
updated MANIFEST 
moved columnate to from dir.c to util.c 
moved more out of dir.c 
removed a few syntax errors 
broken view.c into different files 
renamed view_left view_hscroll 
moved most of selection_fd to text_fd 
more from view.c to util.c 
removed selection_fd altogether 
Kill patch from Steve_Kilbane 
replaced view_isvisible with ISVISIBLE 
compiles with view.c all split up 
renamed the view-related files 
renamed the view-related files 
0.13.19 
scrollbugfix from Steve_Kilbane 
removed text.h from view.c 
tinkering with vgeom 
moved stuff from vsearch to text2 
moved b3 from mouse.c to vsearch.c 
view_goto simplified 
lotsa internal changes 
rename data_labelmodified to data_setlabel 
rename data_backall to data_backupall 
broke search.c into click.c and sambuf.c 
removed text_clear 
removed text_delete 
split and tidied text2.c 
uses sam regexp 
0.13.20 
remove garyfind hack
bugfix undo.c (was causing core dumps)
msg_new now matches doco
0.13.21
exec dies if cannot chdir 
tile patch from Bill Trost
0.13.22
utfwrite handles write returning
_all_ the resize boxes of a window are updated when the window is written
right undo stuff happens for directory read
misc. patches from Bill Trost
realpath uses cached value of cwd
tag_init no longer uses fgets
0.13.23

selected window display patch -- Bill Trost 
directory speedup patch -- Bill Trost 
aesthetic improvements patch -- awatson@NMSU.Edu 
made Alan's and Bill's aesthetics patches cooperate a bit 
noutput bugfix -- Bill Trost, Alan Watson 
0.13.24

make sure anchor is at end of tag for new windows 
organising builtins
bugfix: scrollbar geometry a bit wierd
separated out all operations on lines into line.c 
scrolling backwards through a file now much faster
text_linenumber simplified 
tute/start updated a bit
0.13.25
stop complaining if cannot open .wilytools -- awatson@NMSU.Edu
prettier scrolling, wily -e -- awatson@NMSU.Edu 
text_replace in tag doesn't get forwarded 
fix the problems with switching between Fonts -- Bill Trost 
append WMAINTAG, rather than replace it -- Steve Kotsopoulos 
undo patch --Bill
fixes to above --  Gary 
getname, gettools -- gary
0.13.26


wily/win.c
/*******************************************
 *	add and delete windows
 *******************************************/

#include "tile.h"
static void	win_place(Tile *col, Text *tag, Text *body);

/* Free the resources tied up by 'win'.  Return 0 for success. */
int
win_del(Tile *w) {
	if(!w)
		return 0;

	assert(ISWIN(w));

	/* make sure we can delete the body */
	if(view_delete(w->body)){
		return -1;
	}
	/* delete the tag, and the tile */
	view_delete(w->tag);
	tile_del(w);
	return 0;
}

/* Return the window associated with 'tile', or 0. */
Tile*
tile_win(Tile*tile) {
	return (tile && tile->body) ? tile : 0;
}

/* clone 'win' */
void
win_clone(Tile *win) {
	Text *tag, *body;
	
	assert(ISWIN(win));
	
	tag = view_text(win->tag);
	body = view_text(win->body);
	win_place(win->up, tag, body);
}

/* Create a window to represent 'path' */
void
win_new(char*path, Text*tag, Text*body) {
	Tile * col;
	
	col = findcol(path);	
	win_place(col, tag, body);
}

/* Add some text to w's tag representing the current selection */
void
win_anchor(Tile*w, char *arg) {
	char	buf[80];

	assert(ISWIN(w));

	view_getdot(w->body, buf, arg!=0);
	tag_addtool(view_text(w->tag), buf);
}

/** Create and place a new window with the given 'tag' and 'body'
 * text, in the given 'col'umn
 **/
static void
win_place(Tile *col, Text *tag, Text *body) {
	Tile	*win;
	int	max, min;

	findplace(col, &min, &max);
	win = tile_new(V,  min, max, tagheight, col, tag, body);
	list_add(col, win);
	assert(ISWIN(win));
}

